loader: attach programs using tcx

rgo3 opened the pull request:
For more detailed descriptions, please refer to the individual commits.

On a high level, this PR:

- attaches TC progs using bpf_link (tcx) respecting upgrade and downgrade paths
- uses per-endpoint bpffs dirs, at e.g. /sys/fs/bpf/cilium/endpoints/12345/links/cil_to_container
- adds netns-driven tests for attaching skb progs via tcx
- adds a Helm flag (enableTXC) to optionally disable the feature to ease integration with other tools

Closes https://github.com/cilium/cilium/issues/27632

rgo3 commented on pkg/datapath/loader/netlink.go:
```diff
@@ -267,45 +272,167 @@ func resolveAndInsertCalls(coll *ebpf.Collection, mapName string, calls []ebpf.M
+
+// hasCiliumTCFilters checks if cilium attached bpf filters to a netlink device.
+// Direction is passed as netlink.HANDLE_MIN_{INGRESS,EGRESS} via parent.
+func hasCiliumTCFilters(ifName string, parent uint32) (bool, error) {
```
idea: should this query for a specific program name instead?

rgo3 commented:
Connectivity tests fail because currently the TCX API in cilium/ebpf doesn't do a feature test, so on kernels without TCX support creating links fails and we don't fall back to the old behavior. Opened https://github.com/cilium/ebpf/pull/1294 to add a feature test in ebpf-go.

The build every commit action will continue to fail because I've separate commits for vendoring the lib changes and fixing the query programs API breakage. I'll keep them separate so the fix can easily be applied to the renovate PR that updates the ebpf-go dependency.

rgo3 commented:
/test

rgo3 commented:
/test

ti-mo commented:
/test

ti-mo commented:
Needs rebase on top of https://github.com/cilium/cilium/pull/31553 when merged.

rgo3 commented:
ACK for Timos refactorings. Can't actually give the PR real approval, because it's my own PR :grin:

ti-mo commented:
/test

ti-mo commented:
/test

rgo3 commented:
/test

rgo3 commented:
/test

borkmann commented on pkg/datapath/loader/netlink.go:
```diff
@@ -308,19 +308,14 @@ func attachTCProgram(device netlink.Link, prog *ebpf.Program, progName, bpffsDir
+		err = removeTCFilters(device, parent)
+		if err != nil {
+			log.WithError(err).Error("Unable to remove legacy TC attachments")
+		}
```
Looks good to me! Thanks Robin!

rgo3 commented:
/test

rgo3 committed:
loader: infrastructure for attaching skb programs using the tcx API

This commit adds the necessary infrastructure to attach bpf programs operating
on sk_buff using the kernel's new tcx hook.

Enabling the functionality in the agent's endpoint attachment path happens in
a follow-up commit.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

borkmann committed:
bpf, tcx: Clear tc_classid field

This is needed for tcx given it does not automatically clear the
tc_classid cb field and could contain garbage from upper layers
of the stack. This later maps to skb->tc_index and given in Cilium
code we utilize it, we should explicitly zero the field like we
do with other cb buffers.

Under tcx and endpoint routes, the test below breaks if the field
if not cleared:

  ./cilium-cli connectivity test --test client-ingress

After the fix the test passes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ti-mo committed:
Helm: add bpf.enableTCX Helm value, enable by default

This commit adds the 'bpf.enableTCX' Helm value to allow disabling tcx
attachments if external tooling integrating with Cilium hasn't caught up
yet, as attaching a tcx program to an interface disables the legacy tc
ingress/egress pipelines.

The agent upgrades and downgrades interfaces seamlessly based on tcx being
enabled or not, so any existing workloads are migrated automatically at
runtime if the config flag is changed and the agent restarted. Rebooting
the node is not necessary.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: wire up tcx attachment logic behind --enable-tcx agent flag

This commit puts the tcx logic in the endpoint attachment path and gates it
behind a new --enable-tcx agent flag. A follow-up commit will use the flag
in the Helm charts' configmap.

attachSKBProgram() now takes a bool to indicate if the user has requested tcx
attachments and seamlessly migrates programs between tcx and legacy attachment
modes in both directions. Of course, this process is contingent on no other
tcx programs being attached to the interface, as that disables legacy tc
execution.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo commented:
/test

ti-mo commented:
/test

ti-mo commented:
/test

rgo3 commented on pkg/datapath/loader/netlink.go:
```diff
@@ -828,60 +748,43 @@ func renameDevice(from, to string) error {
+//
+// If checkEgress is true, returns true if there's both an ingress and
+// egress program attached.
+func (l *loader) DeviceHasTCProgramLoaded(device string, checkEgress bool) (bool, error) {
```
Nit: Should this be DeviceHasSKBProgramLoaded for consistency?

rgo3 commented on pkg/datapath/loader/tc_test.go:
```diff
@@ -0,0 +1,161 @@
+	return p
+}
+
+func TestAttachDetachSKBProgram(t *testing.T) {
```
Should this test be run with tcx disabled? We already cover all the TCX cases with other tests, so this one should maybe focus on legacy tc attach/detach.

ti-mo replied:
Good point, I think I had this in mind but it fell through. I ended up renaming this to `TestAttachDetachSKBProgramLegacy` and adding a separate `TestAttachDetachTCProgram` that only uses the legacy tc APIs. :+1:

rgo3 commented on pkg/datapath/loader/tc_test.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// Upgrade a legacy tc program to tcx.
+func TestAttachSKBUpgrade(t *testing.T) {
```
This test should be skipped if TCX is not supported, similar to the other tests in tcx_test.go.

rgo3 commented on pkg/datapath/loader/tc_test.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// Downgrade a tcx program to legacy tc.
+func TestAttachSKBDowngrade(t *testing.T) {
```
skippidyskip?

rgo3 reviewed (commented):
Overall LGTM, just a couple of nits. Wrt docs, should we mention/explain that enabling tcx will skip any other existing legacy tc attachments?

borkmann commented:
> Overall LGTM, just a couple of nits. Wrt docs, should we mention/explain that enabling tcx will skip any other existing legacy tc attachments?

borkmann committed:
cilium, api: Extend API with datapath attach mode

Extend the agent API to indicate whether Cilium is actually using tcx or
relying on legacy tc so that this can be displayed in `cilium status`.

Status when tcx is active:

  # kubectl exec cilium-4m7nq -- cilium-dbg status
  [...]
  BandwidthManager:        Disabled
  Routing:                 Network: Tunnel [geneve]   Host: Legacy
  Attach Mode:             TCX
  Masquerading:            IPTables [IPv4: Enabled, IPv6: Enabled]
  [...]

Status when inactive:

  # kubectl exec cilium-4m7nq -- cilium-dbg status
  [...]
  BandwidthManager:        Disabled
  Routing:                 Network: Tunnel [geneve]   Host: Legacy
  Attach Mode:             Legacy TC
  Masquerading:            IPTables [IPv4: Enabled, IPv6: Enabled]
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ti-mo commented:
/test

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1803,6 +1803,8 @@ enableMasqueradeRouteSource: false
 # -- Enables IPv6 BIG TCP support which increases maximum IPv6 GSO/GRO limits for nodes and pods
 enableIPv6BIGTCP: false
+# -- Attach endpoint programs using tcx instead of legacy tc hooks on supported kernels.
+enableTCX: true
```
nit: Would it make sense to put this under the `bpf` map, i.e. make the value `bpf.enableTCX`?

ti-mo replied:
@gandro Done, and I simplified the chart logic a bit (always rendering `enable-tcx` to the configmap), PTAL.

gandro replied:
Looks good, thanks! Yeah, our charts are super inconsistent unfortunately

borkmann replied:
At some point probably makes sense to bite the bullet and clean it up for more consistency where feasible. Anyway, moving under bpf makes sense. Lgtm.

gandro reviewed (approved):
My CODEOWNERs look good to me, thanks!

ti-mo commented:
/test

borkmann committed:
cilium, docs: Add note to upgrade guide on tcx

Add a note on tcx for the 1.16 upgrade guide.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab commented on pkg/datapath/loader/loader.go:
```diff
@@ -346,6 +346,13 @@ func removeObsoleteNetdevPrograms(devices []string) error {
+		if err := bpf.Remove(bpffsPath); err != nil {
+			log.WithError(err).WithField(logfields.Device, l.Attrs().Name)
+		}
+
```
Is this chunk a bugfix? Seems we should have always removed this dir or is this the first link there?

borkmann replied:
No, given this is the first time we're using tcx links.

jrfastab commented on pkg/datapath/loader/tcx.go:
```diff
@@ -0,0 +1,173 @@
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UpdateLink(pin, prog)
+	switch {
+	// Update successful, nothing left to do.
```
this is all a bit odd to me. I would either return from all case statements or none. But just a nitpick.

ti-mo replied:
Yeah it does look weird, I'll fix.

jrfastab commented on pkg/datapath/loader/tc.go:
```diff
@@ -4,19 +4,41 @@
+		if err == nil {
+			// Created tcx link, clean up any leftover legacy tc attachments.
+			if err := removeTCFilters(device, parent); err != nil {
+				return fmt.Errorf("legacy tc cleanup after attaching tcx program %s: %w", progName, err)
```
if we throw an error here will upsertTCXProgram leave state changed? Should we revert change to state when returning an error if this is the case?

borkmann replied:
(I will follow-up on this.)

ti-mo replied:
Yeah, let's not bail out if tcx already attached successfully and removing the legacy filter failed somehow.

jrfastab commented:
Some minor comments/questions about the code otherwise lgtm.

learnitall reviewed (approved):
API changes LGTM, thanks!

ti-mo commented:
/test

joestringer commented:
I wonder if this is something we should elevate to `release-note/major` given the potential performance impacts. This seems like a great thing to tout in the release blog.

julianwiedmann commented:
:wave: looking at https://github.com/cilium/cilium/actions/runs/8931592322. What's the current upgrade/downgrade impact of this feature - in particular without https://github.com/cilium/cilium/pull/32228 merged? Do we need to switch some of the `bpf-next` downgrade CI tests to `bpf.enableTCX=false`, until v1.15 is sorted?

borkmann commented:
> 👋 looking at https://github.com/cilium/cilium/actions/runs/8931592322. What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported? Do we need to switch some of the `bpf-next` downgrade CI tests to `bpf.enableTCX=false`, until v1.15 is sorted?

I just did the backport of mentioned PR in https://github.com/cilium/cilium/pull/32337 . The ci-ipsec-upgrade test I've seen sporadically fail on various kernels though, not only bpf-next.

borkmann commented:
> What's the current upgrade/downgrade impact of this feature - in particular without https://github.com/cilium/cilium/pull/32228 backported?

Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.

julianwiedmann commented:
> > What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported?
> 
> Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.

I'm at attempt 7 for the `bpf-next` downgrade from main to v1.15 in https://github.com/cilium/cilium/actions/runs/8927670014, which includes the txc work. And as the workflow is *meant* to be sensitive enough for such tiny blips, I'm very inclined to blame this race window ;).

Can we please flip the default to `bpf.enableTCX=false`, until the next v1.15 patch release is through and we can offer a smooth downgrade?

borkmann commented:
> > > What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported?
> > 
> > 
> > Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.
> 
> I'm at attempt 7 for the `bpf-next` downgrade from `main` to `v1.15` in https://github.com/cilium/cilium/actions/runs/8927670014, which includes the txc work. And as the workflow is _meant_ to be sensitive enough for such tiny blips, I'm very inclined to blame this race window ;).
> 
> Can we please flip the default to `bpf.enableTCX=false`, until the next v1.15 patch release is through and we can offer a smooth downgrade?

Ah yes, I thought the downgrade tests would test against latest v1.15 branch, not patch release. Cc @ti-mo lets default to false until then wdyt?

borkmann commented:
(PR is here https://github.com/cilium/cilium/pull/32342 )

rgo3 commented:
@julianwiedmann I've retriggered the job you referenced previously and it is now passing: https://github.com/cilium/cilium/actions/runs/8927670014/job/24559664765. Your 7 runs prior to that pulled a `v1.15` image before we merged the seamless downgrade path 4h ago. So I don't think `bpf.enableTCX=false` is necessary anymore.

julianwiedmann commented:
> @julianwiedmann I've retriggered the job you referenced previously and it is now passing: https://github.com/cilium/cilium/actions/runs/8927670014/job/24559664765. Your 7 runs prior to that pulled a `v1.15` image before we merged the seamless downgrade path 4h ago. So I don't think `bpf.enableTCX=false` is necessary anymore. I'll also rerun it a couple times more to gather some additional data points.

:shrug: I give up on understanding the downgrade tests. So the **minor** downgrade *is* running against the previous minor's branch tip, and **not** the previous minor's current release. And from a quick git archaeology, that was always the case? Then we should be good indeed. Sorry for the noise. Thank you @rgo3 for keeping me honest!

rgo3 commented:
I feel the same :sweat_smile:. 

But yes, I believe and have seen in the CI logs that we install previous minor's branch tip, in this case `v1.15`. To me that also makes the most sense for the minor tests, given the previous minor's branch tip will be the code we support minor downgrades to once released. Otherwise, we would have to backport things, wait until it's merged and released, and only then can move forward with the PR for `main` (something we did in the past, including this PR), which always felt quite tedious to me.
